### PR TITLE
chore(query): support spill by max memory

### PR DESCRIPTION
**Pull request by zhang2014**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

chore(query): support spill by max memory

- Closes #issue

### CODE REVIEW

**Comment by zhang2014** [comment]
This will enabled spill by default, we will merge after fix cluster mode spill hang.

**Comment by BohuTANG** [comment]
Dev / linux / test_unit (pull_request) failed.

**Comment by zhang2014** [comment]
```
2023-07-18T05:13:42.928770Z ERROR common_tracing::panic_hook: panicked at 'assertion failed: `(left == right)`
  left: `["+----------+----------+", "| Column 0 | Column 1 |", "+----------+----------+", "| 1        | '\u{2}\u{1}'       |", "| 2        | '\u{6}\u{1}'       |", "+----------+----------+"]`,
 right: `["+----------+----------+", "| Column 0 | Column 1 |", "+----------+----------+", "| 2        | '\u{6}\u{1}'       |", "+----------+----------+"]`: "refresh index again"

expected:

[
    "+----------+----------+",
    "| Column 0 | Column 1 |",
    "+----------+----------+",
    "| 1        | '\u{2}\u{1}'       |",
    "| 2        | '\u{6}\u{1}'       |",
    "+----------+----------+",
]
actual:

[
    "+----------+----------+",
    "| Column 0 | Column 1 |",
    "+----------+----------+",
    "| 2        | '\u{6}\u{1}'       |",
    "+----------+----------+",
]

', src/query/expression/src/utils/block_debug.rs:107:5, backtrace: Backtrace [{ fn: "trace", file: "/rustc/8b35c0bb0f833c0077dc57006eb317edde2a2d1e/library/std/src/../../backtrace/src/backtrace/libunwind.rs", line: 93 }, { fn: "trace_unsynchronized<std::backtrace::{impl#4}::create::{closure_env#0}>", file: "/rustc/8b35c0bb0f833c0077dc57006eb317edde2a2d1e/library/std/src/../../backtrace/src/backtrace/mod.rs", line: 66 }, { fn: "create", file: "/rustc/8b35c0bb0f833c0077dc57006eb317edde2a2d1e/library/std/src/backtrace.rs", line: 332 }, { fn: "common_tracing::panic_hook::log_panic" }, { fn: "common_tracing::panic_hook::set_panic_hook::{{closure}}" }, { fn: "call<(&core::panic::panic_info::PanicInfo), (dyn core::ops::function::Fn<(&core::panic::panic_info::PanicInfo), Output=()> + core::marker::Send + core::marker::Sync), alloc::alloc::Global>", file: "/rustc/8b35c0bb0f833c0077dc57006eb317edde2a2d1e/library/alloc/src/boxed.rs", line: 1999 }, { fn: "rust_panic_with_hook", file: "/rustc/8b35c0bb0f833c0077dc57006eb317edde2a2d1e/library/std/src/panicking.rs", line: 709 }, { fn: "{closure#0}", file: "/rustc/8b35c0bb0f833c0077dc57006eb317edde2a2d1e/library/std/src/panicking.rs", line: 597 }, { fn: "__rust_end_short_backtrace<std::panicking::begin_panic_handler::{closure_env#0}, !>", file: "/rustc/8b35c0bb0f833c0077dc57006eb317edde2a2d1e/library/std/src/sys_common/backtrace.rs", line: 151 }, { fn: "begin_panic_handler", file: "/rustc/8b35c0bb0f833c0077dc57006eb317edde2a2d1e/library/std/src/panicking.rs", line: 593 }, { fn: "panic_fmt", file: "/rustc/8b35c0bb0f833c0077dc57006eb317edde2a2d1e/library/core/src/panicking.rs", line: 67 }, { fn: "assert_failed_inner", file: "/rustc/8b35c0bb0f833c0077dc57006eb317edde2a2d1e/library/core/src/panicking.rs" }, { fn: "core::panicking::assert_failed" }, { fn: "common_expression::utils::block_debug::assert_two_blocks_sorted_eq_with_name" }, { fn: "it::aggregating_index::index_refresh::test_index_refresh::{{closure}}" }, { fn: "<core::pin::Pin<P> as core::future::future::Future>::poll" }, { fn: "tokio::runtime::park::CachedParkThread::block_on::{{closure}}" }, { fn: "tokio::runtime::park::CachedParkThread::block_on" }, { fn: "tokio::runtime::context::BlockingRegionGuard::block_on" }, { fn: "tokio::runtime::scheduler::multi_thread::MultiThread::block_on" }, { fn: "tokio::runtime::runtime::Runtime::block_on" }, { fn: "it::aggregating_index::index_refresh::test_index_refresh" }, { fn: "it::aggregating_index::index_refresh::test_index_refresh::{{closure}}" }, { fn: "core::ops::function::FnOnce::call_once" }, { fn: "call_once<fn() -> core::result::Result<(), alloc::string::String>, ()>", file: "/rustc/8b35c0bb0f833c0077dc57006eb317edde2a2d1e/library/core/src/ops/function.rs", line: 250 }, { fn: "__rust_begin_short_backtrace<core::result::Result<(), alloc::string::String>, fn() -> core::result::Result<(), alloc::string::String>>", file: "/rustc/8b35c0bb0f833c0077dc57006eb317edde2a2d1e/library/test/src/lib.rs", line: 655 }, { fn: "call_once<(), (dyn core::ops::function::FnOnce<(), Output=core::result::Result<(), alloc::string::String>> + core::marker::Send), alloc::alloc::Global>", file: "/rustc/8b35c0bb0f833c0077dc57006eb317edde2a2d1e/library/alloc/src/boxed.rs", line: 1985 }, { fn: "call_once<core::result::Result<(), alloc::string::String>, alloc::boxed::Box<(dyn core::ops::function::FnOnce<(), Output=core::result::Result<(), alloc::string::String>> + core::marker::Send), alloc::alloc::Global>>", file: "/rustc/8b35c0bb0f833c0077dc57006eb317edde2a2d1e/library/core/src/panic/unwind_safe.rs", line: 271 }, { fn: "do_call<core::panic::unwind_safe::AssertUnwindSafe<alloc::boxed::Box<(dyn core::ops::function::FnOnce<(), Output=core::result::Result<(), alloc::string::String>> + core::marker::Send), alloc::alloc::Global>>, core::result::Result<(), alloc::string::String>>", file: "/rustc/8b35c0bb0f833c0077dc57006eb317edde2a2d1e/library/std/src/panicking.rs", line: 500 }, { fn: "try<core::result::Result<(), alloc::string::String>, core::panic::unwind_safe::AssertUnwindSafe<alloc::boxed::Box<(dyn core::ops::function::FnOnce<(), Output=core::result::Result<(), alloc::string::String>> + core::marker::Send), alloc::alloc::Global>>>", file: "/rustc/8b35c0bb0f833c0077dc57006eb317edde2a2d1e/library/std/src/panicking.rs", line: 464 }, { fn: "catch_unwind<core::panic::unwind_safe::AssertUnwindSafe<alloc::boxed::Box<(dyn core::ops::function::FnOnce<(), Output=core::result::Result<(), alloc::string::String>> + core::marker::Send), alloc::alloc::Global>>, core::result::Result<(), alloc::string::String>>", file: "/rustc/8b35c0bb0f833c0077dc57006eb317edde2a2d1e/library/std/src/panic.rs", line: 142 }, { fn: "run_test_in_process", file: "/rustc/8b35c0bb0f833c0077dc57006eb317edde2a2d1e/library/test/src/lib.rs", line: 678 }, { fn: "{closure#0}", file: "/rustc/8b35c0bb0f833c0077dc57006eb317edde2a2d1e/library/test/src/lib.rs", line: 572 }, { fn: "{closure#1}", file: "/rustc/8b35c0bb0f833c0077dc57006eb317edde2a2d1e/library/test/src/lib.rs", line: 599 }, { fn: "__rust_begin_short_backtrace<test::run_test::run_test_inner::{closure_env#1}, ()>", file: "/rustc/8b35c0bb0f833c0077dc57006eb317edde2a2d1e/library/std/src/sys_common/backtrace.rs", line: 135 }, { fn: "{closure#0}<test::run_test::run_test_inner::{closure_env#1}, ()>", file: "/rustc/8b35c0bb0f833c0077dc57006eb317edde2a2d1e/library/std/src/thread/mod.rs", line: 529 }, { fn: "call_once<(), std::thread::{impl#0}::spawn_unchecked_::{closure#1}::{closure_env#0}<test::run_test::run_test_inner::{closure_env#1}, ()>>", file: "/rustc/8b35c0bb0f833c0077dc57006eb317edde2a2d1e/library/core/src/panic/unwind_safe.rs", line: 271 }, { fn: "do_call<core::panic::unwind_safe::AssertUnwindSafe<std::thread::{impl#0}::spawn_unchecked_::{closure#1}::{closure_env#0}<test::run_test::run_test_inner::{closure_env#1}, ()>>, ()>", file: "/rustc/8b35c0bb0f833c0077dc57006eb317edde2a2d1e/library/std/src/panicking.rs", line: 500 }, { fn: "try<(), core::panic::unwind_safe::AssertUnwindSafe<std::thread::{impl#0}::spawn_unchecked_::{closure#1}::{closure_env#0}<test::run_test::run_test_inner::{closure_env#1}, ()>>>", file: "/rustc/8b35c0bb0f833c0077dc57006eb317edde2a2d1e/library/std/src/panicking.rs", line: 464 }, { fn: "catch_unwind<core::panic::unwind_safe::AssertUnwindSafe<std::thread::{impl#0}::spawn_unchecked_::{closure#1}::{closure_env#0}<test::run_test::run_test_inner::{closure_env#1}, ()>>, ()>", file: "/rustc/8b35c0bb0f833c0077dc57006eb317edde2a2d1e/library/std/src/panic.rs", line: 142 }, { fn: "{closure#1}<test::run_test::run_test_inner::{closure_env#1}, ()>", file: "/rustc/8b35c0bb0f833c0077dc57006eb317edde2a2d1e/library/std/src/thread/mod.rs", line: 528 }, { fn: "call_once<std::thread::{impl#0}::spawn_unchecked_::{closure_env#1}<test::run_test::run_test_inner::{closure_env#1}, ()>, ()>", file: "/rustc/8b35c0bb0f833c0077dc57006eb317edde2a2d1e/library/core/src/ops/function.rs", line: 250 }, { fn: "call_once<(), dyn core::ops::function::FnOnce<(), Output=()>, alloc::alloc::Global>", file: "/rustc/8b35c0bb0f833c0077dc57006eb317edde2a2d1e/library/alloc/src/boxed.rs", line: 1985 }, { fn: "call_once<(), alloc::boxed::Box<dyn core::ops::function::FnOnce<(), Output=()>, alloc::alloc::Global>, alloc::alloc::Global>", file: "/rustc/8b35c0bb0f833c0077dc57006eb317edde2a2d1e/library/alloc/src/boxed.rs", line: 1985 }, { fn: "thread_start", file: "/rustc/8b35c0bb0f833c0077dc57006eb317edde2a2d1e/library/std/src/sys/unix/thread.rs", line: 108 }, { fn: "__pthread_deallocate" }], panic.file: "src/query/expression/src/utils/block_debug.rs", panic.line: 107, panic.column: 5
    at src/common/tracing/src/panic_hook.rs:42
```

 unexpect refresh index result if enable spill. CC: @ariesdevil

**Comment by zhang2014** [comment]
fix unit test failure with https://github.com/datafuselabs/databend/pull/12028/commits/678b344793132a5e5ab832d4a2d0460e1e2be662. CC: @ariesdevil.

File: src/query/service/src/pipelines/processors/transforms/aggregator/transform_aggregate_partial.rs [hunk 1]
```diff
@@ -295,7 +302,8 @@ impl<Method: HashMethodBounds> AccumulatingTransform for TransformPartialAggrega
         if Method::SUPPORT_PARTITIONED {
             if matches!(&self.hash_table, HashTable::HashTable(cell)
                 if cell.len() >= self.settings.convert_threshold ||
-                    cell.allocated_bytes() >= self.settings.spilling_bytes_threshold_per_proc
+                    cell.allocated_bytes() >= self.settings.spilling_bytes_threshold_per_proc ||
```
Comment (BohuTANG) [inline src/query/service/src/pipelines/processors/transforms/aggregator/transform_aggregate_partial.rs:305]: `spilling_bytes_threshold_per_proc` default is 0, so the check:
```
cell.allocated_bytes() >= self.settings.spilling_bytes_threshold_per_proc
```
always true?

Comment (zhang2014) [reply src/query/service/src/pipelines/processors/transforms/aggregator/transform_aggregate_partial.rs:305]: We set spilling_bytes_threshold_per_proc = Usize::Max if spilling_bytes_threshold_per_proc == 0;

**Review by BohuTANG — APPROVED** [review]

**Pull request merged**